### DISCUSSION TITLE
Twin + Goober

### DISCUSSION THREAD

**Discussion by pix2D** [ben-rogerson/twin.macro#366] [Ideas]
https://github.com/cristianbote/goober

Perhaps this would be a good alternative to styled-components / emotion? The bundle size and performance difference sounds very tempting... not sure what is missing in terms of features to achieve that (if anything).

**Comment by ben-rogerson** [comment]
Hey Dan

Have you seen the example with [Goober + Preact](https://github.com/ben-rogerson/twin.examples/tree/master/preact-goober)?

I'm a big fan of Goober and really keen on adding more.

How about I add these examples to the lineup next?

- Next + Goober (TypeScript)
- Gatsby + Goober (TypeScript)

**Reply by pix2D** [reply]
Oh duh, I forgot to check if there's existing examples before typing this. The Goober + Preact example looks great! I will give it a spin a bit later.

Maybe you could add a TS example? I would love a Next + Goober (TS) example as well.

**Reply by ben-rogerson** [reply]
I'm aiming for typescript everything from now on 👍 
I've found it’s much less work to remove typescript from a project than to add it 😆

**Reply by pix2D** [reply]
So I've tried to get it working in Next with TS but at the moment I'm experience the same issue as here https://github.com/ben-rogerson/twin.macro/issues/383. Not sure if it's related to that or not (just the same `object: Object` in the CSS--also in the HTML itself it shows `css="[object Object]"`.

Goober itself seems to work, but any use of `tw` whether in a `styled.el` or in the HTML does not.

**Reply by ben-rogerson** [reply]
Hey dan, I think it might be down to goober support for the interpolation, I've posted some workarounds in [that issue](https://github.com/ben-rogerson/twin.macro/issues/383) 👍

**Reply by pix2D** [reply]
Hey Ben, thanks but I'm not sure that's my issue because just a normal `tw=""` in the HTML didn't work either. I had to revert back to styled-components for now (somehow that _always_ works, funny how that goes 🥲), I'll investigate this again when there's a full Next + Goober + TS example. Thanks for your continued hard work on all this! 👍

**Reply by ben-rogerson** [reply]
I've just added the tw/css props to the Goober example using [@agney/babel-plugin-goober-css-prop](https://goober.js.org/integrations/css-prop). That goober + next + typescript example is coming soon 👍

**Reply by pix2D** [reply]
So with webpack4 adding the above made everything work now! 🎉 With webpack5 it also does work (styling is applied and so on), but in the browser (only) there's the following error:

```
Uncaught TypeError: i is not a function
    l goober.js:1
    React 17
        renderWithHooks
        mountIndeterminateComponent
        beginWork
        callCallback
        invokeGuardedCallbackDev
        invokeGuardedCallback
        beginWork$1
        performUnitOfWork
        workLoopSync
        renderRootSync
        performSyncWorkOnRoot
        scheduleUpdateOnFiber
        updateContainer
        legacyRenderSubtreeIntoContainer
        unbatchedUpdates
        legacyRenderSubtreeIntoContainer
        hydrate
    renderReactElement index.tsx:530
    doRender index.tsx:800
    _callee2$ index.tsx:425
    Babel 8
        tryCatch
        invoke
        defineIteratorMethods
        asyncGeneratorStep
        _next
        _asyncToGenerator
        _asyncToGenerator
        _render
    render index.js:495
    <anonymous> next-dev.js:85
    displayContent fouc.js:14
    displayContent fouc.js:5
    <anonymous> next-dev.js:84
    promise callback* next-dev.js:31
    NextJS 7
        js
        factory
        __webpack_require__
        __webpack_exec__
        <anonymous>
        webpackJsonpCallback
        <anonymous>
```

**Reply by ben-rogerson** [reply]
hey dan, sorrry I missed this - cheers for the heads up, I'll find a fix/workaround in the "Next + Goober (TypeScript)" example
